File: java/src/main/java/io/github/muthuishere/toolnexus/Loop.java
package io.github.muthuishere.toolnexus;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * Loop — a live execution of an Agent, and the completion gate that stops it claiming
 * {@code done} too early. A layer over the shipped §8 client: nothing here changes existing
 * behaviour.
 *
 * <p>The placement law this encodes:
 * <pre>
 *   AgentSpec (the harness) answers "MAY it?"     — capability, ceilings.  Per problem.
 *   RunOptions              answers "with WHAT?"  — model for this call.   Per call.
 *   Loop                    answers "DID it?"     — status, turns.         Observed.
 *   none of them            answers "is it RIGHT?"— a tool, skill or agent.
 * </pre>
 * So {@code Loop} takes no options: it is read, not configured.
 *
 * <p><b>Why this class lives in the root package rather than {@code .agents}:</b>
 * {@link LlmClient.RunResult} is immutable and its constructors are package-private, and the gate
 * must produce a derived result (the accumulated tool calls; the {@code incomplete}/{@code limit}
 * stop). Living here lets it rebuild one WITHOUT widening {@code RunResult}'s public surface —
 * which would be a per-port API divergence for a purely internal need.
 */
public final class Loop {

    /**
     * A POLICY check on a tool call — "may it?", never "is it right?". Return {@code "allow"} (or
     * null) to permit; any other string DENIES with that reason.
     */
    @FunctionalInterface
    public interface Guardrail {
        String check(LlmClient.BeforeToolEvent ev);
    }

    /** What a completion verifier returns. */
    public static final class Verdict {
        public final boolean ok;
        public final String reason;

        public Verdict(boolean ok, String reason) {
            this.ok = ok;
            this.reason = reason == null ? "" : reason;
        }

        public static Verdict pass() { return new Verdict(true, ""); }
        public static Verdict fail(String reason) { return new Verdict(false, reason); }
    }

    /**
     * The gate that stops an agent claiming {@code done} before its work verifies. It lives on the
     * agent spec, so it TRAVELS with the agent through delegation — which a host-side retry loop
     * cannot do.
     */
    public static final class Completion {
        /** Judges the run. Receives the tool calls ACCUMULATED across attempts. */
        public final Function<LlmClient.RunResult, Verdict> verify;
        /** REQUIRED. An unbounded verify loop is a denial-of-service on the caller's bill. */
        public final int maxAttempts;

        public Completion(Function<LlmClient.RunResult, Verdict> verify, int maxAttempts) {
            this.verify = verify;
            this.maxAttempts = maxAttempts;
        }
    }

    /**
     * What varies PER CALL. {@code model} is here — not on the spec's default and not on the
     * loop — so the same conversation may change model between turns.
     */
    public static final class RunOptions {
        /** Overrides the agent's model for this call only. Null ⇒ the agent's. */
        public String model;

        public RunOptions model(String v) { this.model = v; return this; }
    }

    /**
     * What a run reports. {@code status} reuses the SHIPPED vocabulary — no new status strings are
     * minted (SPEC.md pins TaskStatus identical across ports).
     */
    public static final class Outcome {
        public final String text;
        public final String status;      // done | incomplete | pending | error
        /** Named whenever {@code status} is not {@code done} — never a silent stop. */
        public final String stoppedBy;
        public final int attempts;
        public final int turns;
        public final LlmClient.RunResult result;

        Outcome(String text, String status, String stoppedBy, int attempts, int turns,
                LlmClient.RunResult result) {
            this.text = text;
            this.status = status;
            this.stoppedBy = stoppedBy;
            this.attempts = attempts;
            this.turns = turns;
            this.result = result;
        }
    }

    /** What the gate calls to run one attempt. */
    @FunctionalInterface
    public interface Ask {
        LlmClient.RunResult run(String prompt);
    }

    private final LlmClient.Options options;
    private final Toolkit toolkit;
    private final String soul;
    private final Completion completion;
    private final List<Guardrail> guardrails;
    private final LlmClient.Hooks hooks;
    private int turns;
    private String status = "idle";

    /**
     * Takes client OPTIONS rather than a built client, because a per-call {@code model} override
     * must be able to change the model — which is fixed when a client is constructed.
     */
    public Loop(LlmClient.Options options, Toolkit toolkit, String soul,
                List<Guardrail> guardrails, LlmClient.Hooks hooks, Completion completion) {
        this.options = options;
        this.toolkit = toolkit;
        this.soul = soul;
        this.guardrails = guardrails;
        this.hooks = hooks;
        this.completion = completion;
    }

    /** Observed, never set by the caller. */
    public String status() { return status; }

    /** Model round trips this loop has spent. */
    public int turns() { return turns; }

    public Outcome run(String prompt) { return run(prompt, new RunOptions()); }

    public Outcome run(String prompt, RunOptions opts) {
        LlmClient client = LlmClient.create(clientOptions(opts));
        status = "running";

        int[] attempts = {0};
        List<Object>[] history = new List[]{null};
        Ask ask = text -> {
            attempts[0]++;
            LlmClient.RunResult r = client.run(text, toolkit, history[0]);
            turns += r.turns;
            history[0] = r.messages;
            return r;
        };

        LlmClient.RunResult r;
        try {
            r = runGated(ask, prompt, completion);
        } catch (RuntimeException e) {
            status = "error";
            throw e;
        }

        if (r.status != null && !"done".equals(r.status)) {
            status = r.status;
            String stoppedBy = "completion".equals(r.limit) ? r.text : "run reported " + r.status;
            return new Outcome(r.text, r.status, stoppedBy, attempts[0], turns, r);
        }
        status = "idle";
        return new Outcome(r.text, "done", null, attempts[0], turns, r);
    }

    /**
     * Applies a per-call model override via {@code requestParams} ({@code model} is not in the
     * forbidden set — the client forbids only messages/tools/stream).
     */
    private LlmClient.Options clientOptions(RunOptions opts) {
        LlmClient.Options o = copyOptions(options);
        if (soul != null && !soul.isEmpty() && (o.systemPrompt == null || o.systemPrompt.isEmpty())) {
            o.systemPrompt = soul;
        }
        o.hooks = guardedHooks(guardrails, hooks != null ? hooks : o.hooks);
        if (opts != null && opts.model != null && !opts.model.isEmpty()) {
            Map<String, Object> rp = new java.util.LinkedHashMap<>();
            if (o.requestParams != null) rp.putAll(o.requestParams);
            rp.put("model", opts.model);
            o.requestParams = rp;
        }
        return o;
    }

    /**
     * {@code LlmClient.Options} has no copy method, so mirror the fields — the loop must never
     * mutate the caller's object, since one Options may back several loops.
     */
    private static LlmClient.Options copyOptions(LlmClient.Options o) {
        LlmClient.Options c = new LlmClient.Options();
        c.baseUrl = o.baseUrl; c.style = o.style; c.model = o.model; c.apiKey = o.apiKey;
        c.headers = o.headers; c.systemPrompt = o.systemPrompt; c.maxTurns = o.maxTurns;
        c.hooks = o.hooks; c.retries = o.retries; c.retryBaseMs = o.retryBaseMs;
        c.timeoutMs = o.timeoutMs; c.store = o.store; c.onMetric = o.onMetric;
        c.waitFor = o.waitFor; c.requestParams = o.requestParams;
        c.bodyTransform = o.bodyTransform; c.httpClient = o.httpClient; c.onError = o.onError;
        return c;
    }

    /**
     * Compiles guardrails into one {@code beforeTool} with FIRST-DENY-WINS, composed ahead of any
     * hook already set. No guardrails ⇒ {@code hooks} is returned untouched, so absent is
     * byte-identical.
     */
    public static LlmClient.Hooks guardedHooks(List<Guardrail> guardrails, LlmClient.Hooks hooks) {
        if (guardrails == null || guardrails.isEmpty()) return hooks;
        final Function<LlmClient.BeforeToolEvent, LlmClient.ToolOverride> prior =
                hooks != null ? hooks.beforeTool : null;

        LlmClient.Hooks merged = new LlmClient.Hooks();
        if (hooks != null) {
            merged.beforeLLM = hooks.beforeLLM;
            merged.afterLLM = hooks.afterLLM;
            merged.afterTool = hooks.afterTool;
        }
        merged.beforeTool = ev -> {
            for (Guardrail g : guardrails) {
                String verdict = g.check(ev);
                if (verdict != null && !verdict.isEmpty() && !"allow".equals(verdict)) {
                    return new LlmClient.ToolOverride(
                            null, new ToolResult("denied: " + verdict, true, null));
                }
            }
            return prior != null ? prior.apply(ev) : null;
        };
        return merged;
    }

    /**
     * The built-in completion verifier. Reads the SHIPPED {@code todowrite} builtin's result
     * metadata and requires every item to be checked.
     *
     * <p>Structural, not domain: it counts unchecked boxes and never learns what a todo means, so
     * the loop stays domain-blind. No plan declared ⇒ nothing to verify ⇒ pass, so the gate never
     * punishes an agent that does not use the builtin.
     */
    public static Verdict allTodosDone(LlmClient.RunResult result) {
        for (int i = result.toolCalls.size() - 1; i >= 0; i--) {
            LlmClient.ToolCall call = result.toolCalls.get(i);
            if (!"todowrite".equals(call.name)) continue;
            Object raw = call.metadata == null ? null : call.metadata.get("todos");
            if (!(raw instanceof List<?> todos)) return Verdict.pass();

            List<String> open = new ArrayList<>();
            for (Object item : todos) {
                if (!(item instanceof Map<?, ?> m)) continue;
                if (!Boolean.TRUE.equals(m.get("completed"))) {
                    Object text = m.get("text");
                    open.add(text == null ? "" : String.valueOf(text));
                }
            }
            return open.isEmpty()
                    ? Verdict.pass()
                    : Verdict.fail(open.size() + " item(s) still open: " + String.join("; ", open));
        }
        return Verdict.pass();
    }

    /**
     * Wraps a client run with the completion gate. SHARED by the standalone loop and the §7D
     * runtime turn, so a delegated child gets exactly the same guarantee as a directly-driven one.
     *
     * <p>Rule 2 in force: a run that is {@code pending} (suspended on a human) or otherwise
     * non-done already carries its own reason, so the gate never re-judges it. That keeps
     * {@code pending} and {@code incomplete} distinct — the caller can always tell whether it owes
     * an Answer or a fix.
     */
    public static LlmClient.RunResult runGated(Ask ask, String prompt, Completion completion) {
        if (completion == null) return ask.run(prompt);
        if (completion.maxAttempts < 1) {
            throw new IllegalArgumentException("toolnexus: Completion.maxAttempts must be >= 1");
        }
        if (completion.verify == null) {
            throw new IllegalArgumentException("toolnexus: Completion.verify is required");
        }

        List<LlmClient.ToolCall> accumulated = new ArrayList<>();
        LlmClient.RunResult last = null;
        String reason = "";

        for (int attempt = 1; attempt <= completion.maxAttempts; attempt++) {
            String text = attempt == 1
                    ? prompt
                    : "Your work did not verify: " + reason + ". Fix it and finish.";
            LlmClient.RunResult r = ask.run(text);

            // The gate judges the ACCUMULATED work, so an agent cannot escape it by declining to
            // re-declare its plan on a retry.
            accumulated.addAll(r.toolCalls);
            r = rebuild(r, r.text, new ArrayList<>(accumulated), r.status, r.limit);
            last = r;

            if (r.status != null && !"done".equals(r.status)) {
                // The run stopped for its own reason (suspension, budget). If the gate was
                // mid-retry the caller must learn BOTH — otherwise a budget stop masks the
                // verification failure and they never see why it was looping.
                if (!reason.isEmpty() && !"pending".equals(r.status)) {
                    r = rebuild(r, r.text + " [while verifying: attempt " + attempt
                            + " last failed: " + reason + "]", r.toolCalls, r.status, r.limit);
                }
                return r;
            }

            Verdict verdict = completion.verify.apply(r);
            if (verdict.ok) return r;
            reason = verdict.reason.isEmpty() ? "unspecified" : verdict.reason;
        }

        // Structured, not prose: `limit` is how a caller (and the §7D runtime) tells WHICH limit
        // stopped the run. `text` carries the human reason.
        return rebuild(last,
                "completion.verify failed " + completion.maxAttempts + "x: " + reason,
                last.toolCalls, "incomplete", "completion");
    }

    /** RunResult is immutable, so a "modified" result is a new one carrying the rest verbatim. */
    private static LlmClient.RunResult rebuild(LlmClient.RunResult r, String text,
                                               List<LlmClient.ToolCall> toolCalls,
                                               String status, String limit) {
        return new LlmClient.RunResult(text, r.messages, toolCalls, r.turns, r.usage, r.model,
                status, r.pending, limit);
    }
}

File: csharp/src/Toolnexus/Agents/Loop.cs
using System.Text.Json;

namespace Toolnexus.Agents;

/// <summary>
/// A POLICY check on a tool call — "may it?", never "is it right?". Return
/// <c>"allow"</c> (or null) to permit; any other string DENIES with that reason.
/// </summary>
public delegate string? Guardrail(LlmClient.BeforeToolEvent ev);

/// <summary>What a completion verifier returns.</summary>
public sealed record Verdict(bool Ok, string Reason = "");

/// <summary>
/// The gate that stops an agent claiming <c>done</c> before its work verifies. It lives on the
/// agent spec, so it TRAVELS with the agent through delegation — which a host-side retry loop
/// cannot do.
/// </summary>
public sealed class Completion
{
    /// <summary>Judges the run. Receives the tool calls ACCUMULATED across attempts.</summary>
    public Func<LlmClient.RunResult, Verdict> Verify { get; set; } = _ => new Verdict(true);

    /// <summary>REQUIRED. An unbounded verify loop is a denial-of-service on the caller's bill.</summary>
    public int MaxAttempts { get; set; }
}

/// <summary>What varies PER CALL. <c>Model</c> is here — not on the spec's default and not on the
/// loop — so the same conversation may change model between turns.</summary>
public sealed class LoopRunOptions
{
    /// <summary>Overrides the agent's model for this call only. Null ⇒ the agent's.</summary>
    public string? Model { get; set; }
}

/// <summary>What a run reports. <c>Status</c> reuses the SHIPPED vocabulary — no new status
/// strings are minted (SPEC.md pins TaskStatus identical across ports).</summary>
public sealed record Outcome(
    string Text,
    string Status,
    string? StoppedBy,
    int Attempts,
    int Turns,
    LlmClient.RunResult? Result);

/// <summary>
/// Loop — a live execution of an Agent, and the completion gate. A layer over the shipped §8
/// client: nothing here changes existing behaviour.
///
/// <para>The placement law this encodes:</para>
/// <list type="bullet">
///   <item><description>AgentSpec (the harness) answers "MAY it?" — capability, ceilings. Per problem.</description></item>
///   <item><description>LoopRunOptions answers "with WHAT?" — model for this call. Per call.</description></item>
///   <item><description>Loop answers "DID it?" — status, turns. Observed.</description></item>
///   <item><description>none of them answers "is it RIGHT?" — that is a tool, skill or agent.</description></item>
/// </list>
/// <para>So Loop takes no options: it is read, not configured.</para>
/// </summary>
public sealed class Loop
{
    private readonly Agent _agent;
    private readonly LlmClient.Options _options;
    private readonly Toolkit _toolkit;
    private int _turns;

    /// <param name="options">Client OPTIONS rather than a built client, because a per-call
    /// <c>Model</c> override must be able to change the model — fixed at construction.</param>
    public Loop(Agent agent, LlmClient.Options options, Toolkit toolkit)
    {
        _agent = agent;
        _options = options;
        _toolkit = toolkit;
    }

    /// <summary>Observed, never set by the caller.</summary>
    public string Status { get; private set; } = "idle";

    /// <summary>Model round trips this loop has spent.</summary>
    public int Turns => _turns;

    public async Task<Outcome> RunAsync(string prompt, LoopRunOptions? opts = null)
    {
        var completion = _agent.Spec.Completion;
        var client = LlmClient.Create(ClientOptionsFor(opts));
        Status = "running";

        var attempts = 0;
        List<object>? history = null;

        async Task<LlmClient.RunResult> Ask(string text)
        {
            attempts++;
            var r = await client.RunAsync(text, _toolkit, history).ConfigureAwait(false);
            _turns += r.Turns;
            history = r.Messages;
            return r;
        }

        LlmClient.RunResult result;
        try
        {
            result = await RunGatedAsync(Ask, prompt, completion).ConfigureAwait(false);
        }
        catch
        {
            Status = "error";
            throw;
        }

        if (!string.IsNullOrEmpty(result.Status) && result.Status != "done")
        {
            Status = result.Status;
            var stoppedBy = result.Limit == "completion" ? result.Text : $"run reported {result.Status}";
            return new Outcome(result.Text, result.Status, stoppedBy, attempts, _turns, result);
        }

        Status = "idle";
        return new Outcome(result.Text, "done", null, attempts, _turns, result);
    }

    /// <summary>Applies a per-call model override via <c>RequestParams</c> (<c>model</c> is not in
    /// the forbidden set — the client forbids only messages/tools/stream).</summary>
    private LlmClient.Options ClientOptionsFor(LoopRunOptions? opts)
    {
        var o = CloneOptions(_options);
        if (!string.IsNullOrEmpty(_agent.Spec.Soul) && string.IsNullOrEmpty(o.SystemPrompt))
            o.SystemPrompt = _agent.Spec.Soul;
        o.Hooks = LoopSupport.GuardedHooks(_agent.Spec.Guardrails, _agent.Spec.Hooks ?? o.Hooks);

        if (string.IsNullOrEmpty(opts?.Model)) return o;
        var rp = o.RequestParams is null
            ? new Dictionary<string, object?>()
            : new Dictionary<string, object?>(o.RequestParams);
        rp["model"] = opts!.Model;
        o.RequestParams = rp;
        return o;
    }

    /// <summary>
    /// Wraps a client run with the completion gate. SHARED by the standalone loop and the §7D
    /// runtime turn, so a delegated child gets exactly the same guarantee as a directly-driven one.
    ///
    /// <para>Rule 2 in force: a run that is <c>pending</c> (suspended on a human) or otherwise
    /// non-done already carries its own reason, so the gate never re-judges it. That keeps
    /// <c>pending</c> and <c>incomplete</c> distinct — the caller can always tell whether it owes
    /// an Answer or a fix.</para>
    /// </summary>
    internal static async Task<LlmClient.RunResult> RunGatedAsync(
        Func<string, Task<LlmClient.RunResult>> ask, string prompt, Completion? completion)
    {
        if (completion is null) return await ask(prompt).ConfigureAwait(false);
        if (completion.MaxAttempts < 1)
            throw new ArgumentException("toolnexus: Completion.MaxAttempts must be >= 1");
        if (completion.Verify is null)
            throw new ArgumentException("toolnexus: Completion.Verify is required");

        var accumulated = new List<LlmClient.ToolCall>();
        LlmClient.RunResult? last = null;
        var reason = "";

        for (var attempt = 1; attempt <= completion.MaxAttempts; attempt++)
        {
            var text = attempt == 1 ? prompt : $"Your work did not verify: {reason}. Fix it and finish.";
            var r = await ask(text).ConfigureAwait(false);

            // The gate judges the ACCUMULATED work, so an agent cannot escape it by declining to
            // re-declare its plan on a retry.
            accumulated.AddRange(r.ToolCalls);
            // RunResult is IMMUTABLE in this port, so the judged view is a rebuild rather
            // than a mutation. Same semantics, different idiom.
            r = Rebuild(r, toolCalls: new List<LlmClient.ToolCall>(accumulated));
            last = r;

            if (!string.IsNullOrEmpty(r.Status) && r.Status != "done")
            {
                // The run stopped for its own reason (suspension, budget). If the gate was
                // mid-retry the caller must learn BOTH — otherwise a budget stop masks the
                // verification failure and they never see why it was looping.
                if (reason.Length > 0 && r.Status != "pending")
                    r = Rebuild(r, text: $"{r.Text} [while verifying: attempt {attempt} last failed: {reason}]");
                return r;
            }

            var verdict = completion.Verify(r);
            if (verdict.Ok) return r;
            reason = string.IsNullOrEmpty(verdict.Reason) ? "unspecified" : verdict.Reason;
        }

        // Structured, not prose: `Limit` is how a caller (and the §7D runtime) tells WHICH limit
        // stopped the run. `Text` carries the human reason.
        return Rebuild(last!,
            text: $"completion.verify failed {completion.MaxAttempts}x: {reason}",
            status: "incomplete", limit: "completion");
    }

    /// <summary>LlmClient.Options has no Clone, so copy the fields the loop may change and
    /// carry the rest by reference — the loop never mutates the caller's object.</summary>
    private static LlmClient.Options CloneOptions(LlmClient.Options o) => new()
    {
        BaseUrl = o.BaseUrl, Style = o.Style, Model = o.Model, ApiKey = o.ApiKey,
        Headers = o.Headers, SystemPrompt = o.SystemPrompt, MaxTurns = o.MaxTurns,
        Hooks = o.Hooks, Retries = o.Retries, RetryBaseMs = o.RetryBaseMs, TimeoutMs = o.TimeoutMs,
        Store = o.Store, OnMetric = o.OnMetric, WaitFor = o.WaitFor,
        RequestParams = o.RequestParams, BodyTransform = o.BodyTransform,
        HttpClient = o.HttpClient, HttpHandler = o.HttpHandler, OnError = o.OnError,
    };

    /// <summary>RunResult is immutable here, so a "modified" result is a new one. Everything not
    /// named is carried over verbatim.</summary>
    private static LlmClient.RunResult Rebuild(
        LlmClient.RunResult r,
        string? text = null,
        List<LlmClient.ToolCall>? toolCalls = null,
        string? status = null,
        string? limit = null) =>
        new(text ?? r.Text, r.Messages, toolCalls ?? r.ToolCalls, r.Turns, r.Usage, r.Model,
            status ?? r.Status, r.Pending, limit ?? r.Limit);
}

/// <summary>Helpers shared by the loop, the agent spec projection and the runtime turn.</summary>
public static class LoopSupport
{
    /// <summary>
    /// Compiles guardrails into one <c>BeforeTool</c> with FIRST-DENY-WINS, composed ahead of any
    /// hook already set. No guardrails ⇒ <paramref name="hooks"/> is returned untouched, so absent
    /// is byte-identical.
    /// </summary>
    public static LlmClient.Hooks? GuardedHooks(List<Guardrail>? guardrails, LlmClient.Hooks? hooks)
    {
        if (guardrails is null || guardrails.Count == 0) return hooks;
        var prior = hooks?.BeforeTool;
        var merged = hooks is null
            ? new LlmClient.Hooks()
            : new LlmClient.Hooks
            {
                BeforeLLM = hooks.BeforeLLM, AfterLLM = hooks.AfterLLM,
                BeforeTool = hooks.BeforeTool, AfterTool = hooks.AfterTool,
            };

        merged.BeforeTool = ev =>
        {
            foreach (var rail in guardrails)
            {
                var verdict = rail(ev);
                if (!string.IsNullOrEmpty(verdict) && verdict != "allow")
                    return new LlmClient.ToolOverride
                    {
                        Result = new ToolResult($"denied: {verdict}", true, null),
                    };
            }
            return prior?.Invoke(ev);
        };
        return merged;
    }

    /// <summary>
    /// The built-in completion verifier. Reads the SHIPPED <c>todowrite</c> builtin's result
    /// metadata and requires every item to be checked.
    ///
    /// <para>Structural, not domain: it counts unchecked boxes and never learns what a todo means,
    /// so the loop stays domain-blind. No plan declared ⇒ nothing to verify ⇒ pass, so the gate
    /// never punishes an agent that does not use the builtin.</para>
    /// </summary>
    public static Verdict AllTodosDone(LlmClient.RunResult result)
    {
        for (var i = result.ToolCalls.Count - 1; i >= 0; i--)
        {
            if (result.ToolCalls[i].Name != "todowrite") continue;
            var metadata = result.ToolCalls[i].Metadata;
            if (metadata is null || !metadata.TryGetValue("todos", out var raw)) return new Verdict(true);

            var open = new List<string>();
            if (raw is IEnumerable<object?> items)
            {
                foreach (var item in items)
                {
                    var (completed, text) = ReadTodo(item);
                    if (!completed) open.Add(text);
                }
            }
            else
            {
                return new Verdict(true);
            }

            return open.Count > 0
                ? new Verdict(false, $"{open.Count} item(s) still open: {string.Join("; ", open)}")
                : new Verdict(true);
        }
        return new Verdict(true);
    }

    private static (bool Completed, string Text) ReadTodo(object? item) => item switch
    {
        JsonElement je when je.ValueKind == JsonValueKind.Object => (
            je.TryGetProperty("completed", out var c) && c.ValueKind == JsonValueKind.True,
            je.TryGetProperty("text", out var t) ? t.GetString() ?? "" : ""),
        IDictionary<string, object?> d => (
            d.TryGetValue("completed", out var c) && c is true,
            d.TryGetValue("text", out var t) ? t?.ToString() ?? "" : ""),
        _ => (true, ""),
    };
}
